src/object.jl

```diff
@@ -186,21 +192,45 @@ mutable struct PrefechedDownloadStream{T <: Object} <: IO
         len = parse(Int, HTTP.header(resp, "Content-Length", "0"))
         et = etag(HTTP.header(resp, "ETag", ""))
         object = Object(store, credentials, String(key), Int(len), String(et))
-        return PrefechedDownloadStream(object, prefetch_size; prefetch_multipart_size)
+        return PrefetchedDownloadStream(object, prefetch_size; prefetch_multipart_size)
     end
 end
-Base.eof(io::PrefechedDownloadStream) = io.pos >= io.len
-Base.bytesavailable(io::PrefechedDownloadStream) = io.len - io.pos + 1
-function Base.close(io::PrefechedDownloadStream)
+Base.eof(io::PrefetchedDownloadStream) = io.pos >= io.len
+bytesremaining(io::PrefetchedDownloadStream) = io.len - io.pos + 1
+function Base.bytesavailable(io::PrefetchedDownloadStream)
+    return isnothing(io.buf) ? 0 : bytesavailable(io.buf::PrefetchBuffer)
+end
+function Base.close(io::PrefetchedDownloadStream)
     close(io.prefetch_queue)
     close(io.download_queue)
-    Base.notify_error(io.cond.cond_wait, Base.closed_exception())
+    Base.@lock io.cond.cond_wait begin
+        Base.notify_error(io.cond.cond_wait, Base.closed_exception())
+    end
```

Comment:
This is a bit esoteric. I guess it's probably copied from Base IO objects? Can you maybe add a comment indicating where this came from and why it's needed?